src/backend/distributed/deparser/deparse_role_stmts.c

```diff
@@ -430,13 +438,46 @@ AppendGrantWithAdminOption(StringInfo buf, GrantRoleStmt *stmt)
 		DefElem *opt = NULL;
 		foreach_ptr(opt, stmt->opt)
 		{
-			bool admin_option = false;
-			char *optval = defGetString(opt);
-			if (strcmp(opt->defname, "admin") == 0 &&
-				parse_bool(optval, &admin_option) && admin_option)
+			 if (strcmp(opt->defname, "admin") == 0)
+            {
+                bool admin_option = false;
+				char *optval = defGetString(opt);
+				if (parse_bool(optval, &admin_option) && admin_option)
+				{
+					appendStringInfo(buf, " WITH ADMIN OPTION");
+				}
+			}
+			else if (strcmp(opt->defname, "inherit") == 0)
+            {
+                bool inherit_option = false;
+                char *optval = defGetString(opt);
+                if (parse_bool(optval, &inherit_option))
+                {
+                    if (!inherit_option)
+                    {
+                        appendStringInfo(buf, "WITH INHERIT FALSE");
+                    }
+                    else
+                    {
+                        appendStringInfo(buf, "INHERIT OPTION FOR ");
+                    }
+                }
+            }
+			else if (strcmp(opt->defname, "set") == 0)
 			{
-				appendStringInfo(buf, " WITH ADMIN OPTION");
-				break;
+			    bool set_option = false;
+				char *optval = defGetString(opt);
+				if (parse_bool(optval, &set_option))
+				{
+					if (!set_option)
+					{
+						appendStringInfo(buf, "WITH SET FALSE");
+					}
+					else
+					{
+						appendStringInfo(buf, "SET OPTION FOR ");
+					}
+				}
```

Comment:
As I have mentioned before, you can have multiple options within the same command, and the way it works is, for example:
```SQL
GRANT role1 TO role2 WITH INHERIT FALSE, ADMIN TRUE, SET FALSE;
```
So, the code implementation is still not correct. As I have suggested earlier, given that we will always include INHERIT option, you can start with INHERIT, and then add a comma to the string, and not include "WITH" in the admin and set options. Regression tests will certainly help you with this.
Whereas for revokes we can have just one option (PLEASE DOUBLE CHECK THIS)
```SQL
REVOKE SET OPTION FOR role1 FROM role2;
```